Allow init from objects with __astropy_table__ method

taldcroft opened the pull request:
This is a second try at #4864.

This defines an interface method to allow an arbitrary object to initialize a `Table` subclass if it has an `__astropy_table__` method.

To do:

- [x] Documentation in Table narrative docs
- [x] Changelog

Supercedes and closes #4740.

cc: @astrofrog @TallJimbo @eteq @mhvk

astrofrog commented:
👍

mhvk commented:
I like this better than #4740, but have one suggestion: make all arguments keywords right from the start.  That way it is easier if it becomes needed to extend the amount of information we pass on. It would mean that the docs should be explicit that any implementation should use
```
def __astropy_table__(self, **kwargs):
```
where in principle any keywords that might be of interest could be added (e.g., `__astropy_table__(self, cls=Table, **kwargs)`, but they should have defaults scuh that `object.__astropy_table__()` always returns a `Table` subclass instance.

taldcroft commented:
@mhvk - can I meet you part way and make the required signature as below?
```
def __astropy_table__(self, table_class, copy, **kwargs):
```
The first two are absolutely required of every implementation, and that is a good use case for Python positional arguments.  I really try to avoid pure **kwargs APIs because the docs just don't feel as clear and in this case forces every implementation to start with boilerplate code that is not necessary.

I'm also wondering about your case for `t.__astropy_table__()` to work.  As I understand this method it is purely for astropy `Table` (or subclasses) to call, and that will always provide `table_class` and `copy`.  So making these kwarg arguments with defaults implies a degree of freedom that doesn't exist.  It also implies that the `__astropy_table__` developer needs to actually choose their defaults, and even import `Table`.  Having a signature: `def __astropy_table__(self, cls=Table, ..)` would be especially bad because `Table` would need to be always imported in order to import the module.

I do see your point that if it needs to be future-proofed in case we find at some point that something else is required, in which case we can pass those along as kwargs.

mhvk commented:
Agreed my example is bad. But my future-proofing is partially that I don't know if at some point we wouldn't be happy with a routine passing back *any* astropy-table compatible object, or that `copy` is ignored.

But I'm probably overthinking this, and I do see the big advantage of being able to count on `cls` being passed on: no need to import `Table`, etc. So, let's for sure do `cls`. I think I'd still prefer `copy` being used as a keyword argument, mostly as it often ends up at the end of long lists, but I don't really mind either way.

TallJimbo commented:
As linked above, I've just put the LSST-side implementation out for review.  From that standpoint, all that I *think* remains to be settled here is whether the class argument is `cls`, `table_cls`, `table_class`, etc.

taldcroft commented:
@TallJimbo - since you have already written `cls` and that is a quasi-standard (though usually for class methods), I'll go with `cls` as well.

taldcroft commented:
Minor name change, docs and a changelog on the way and we'll be set.

eteq commented:
FYI, once you've got the docs in feel free to ping me and I'll give the docs and such a final review.

taldcroft commented:
@eteq - I've pushed the advertised changes.  It's going to need a rebase but you should be able to look at the changes now.

taldcroft commented:
(And maybe I'll wait for your review before rebasing to avoid burning more travis time if there is something little that still needs change).

mhvk commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+object ``data`` that provides an ``__astropy_table__`` method.  In this case the
+``__astropy_table__`` method will be called as follows::
+
+  table = data.__astropy_table__(cls, copy)
```
Do add `**kwargs` here, so that at least we can expand later if we have too!
I'd also prefer to have `copy=True` here, so we can change our mind if we have to (but for `cls` I'm now completely convinced...). After all, it doesn't hurt!

mhvk replied:
OK, sorry, I was too quick; you do say so below...

mhvk commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+
+where ``cls`` is the |Table| class or subclass that is being instantiated,
+and ``copy`` indicates whether a copy of the values in ``data`` should be
+provided.  If ``copy`` is ``False`` then the ``__astropy_table__`` method
```
This is very unlike everything I have seen elsewhere: the demand is always the reverse: if `copy=True`, then it cannot be done by-reference. If `False`, it should be done by reference if feasible.

eteq replied:
Agreed - I would change the name to something like ``can_be_copy`` or something similar.

eteq replied:
Also it might be good to add a sentence describing the `True` case - e.g., "If ``copy`` is ``True``, that indicates it is safe for ``__astropy_table__`` to pass in a copy of its data."  (I might be misunderstanding... but whatever the right thing is, put it in there ;) )

mhvk replied:
I think the name `copy` is fine, just the definition here is unusual (and I see no need to guarantee passing references; e.g., for some shape changes and transpositions it is impossible to it by reference).

taldcroft replied:
Interesting.  Maybe astropy.Table has been too conservative in the interpretation.  It definitely always copies when `copy=True` but in at least some circumstances it purposely raises an exception when `copy=False` but it can't return a reference.  Maybe we should make this more liberal into the "return a reference if you can"?  In fact `Table` is somewhat inconsistent about applying the rule.

So I will change the definition accordingly, and probably open an issue to make `Table` use that interpretation.

mhvk replied:
Sounds good -- and a separate issue can also show whether or not I'm overly biased by working with `ndarray`!

TallJimbo replied:
This discussion is exactly the reason I added the `unviewable` argument on lsst/afw#64; it controls what happens when a copy is required for a column, but `copy=False` (the choices are raise an exception, skip the column, or copy the column).  It sounds like it *might* be appropriate for that argument (or an equivalent) to actually be part of the `__init__` here, with the current behavior implying a default of "raise" and the proposal here being to change it to "copy".

mhvk commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+but does illustrate the principles involved.)  Please note the method
+signature::
+
+  def __astropy_table__(self, cls, copy, **kwargs):
```
So, here I would advocate giving `copy` a default value.

taldcroft replied:
I still don't understand why `copy` should be a kwarg with a default.  As far as I can see only adds noise to the API by requiring the developer to pick a default value that *never* gets used.   This method `__astropy_table__` should never ever be called by any code besides the core `Table.__init__` method.  In some sense I want to make the API non-friendly for users so they never get tempted to call it directly instead of doing `Table(my_table)`.

mhvk replied:
OK, agreed.

mhvk commented on astropy/table/table.py:
```diff
@@ -273,6 +273,15 @@ def __init__(self, data=None, masked=None, names=None, dtype=None,
+            # self.__class__ and respects the `copy` arg.  The returned
+            # Table object should NOT then be copied (though the meta
+            # will be deep-copied anyway).
+            data = data.__astropy_table__(self.__class__, copy)
```
I'd prefer this to be `copy=copy` for idiotic safety...

taldcroft replied:
OK, but it seems like the implication is that positional args are being disallowed in astropy internal methods.

mhvk replied:
Ignore this comment...  I think you convinced me it is fine to insist on a particular set.

eteq commented:
(reviewing now)

eteq commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+
+As a simple example, imagine a dict-based table class.  (Note that |Table|
+already can be initialized from a dict-like object, so this is a bit contrived
+but does illustrate the principles involved.)  Please note the method
```
(wordsmithing): double-"note" here, maybe "Pay attention to the method signature"?

eteq commented on CHANGES.rst:
```diff
@@ -118,6 +118,9 @@ New Features
 
   - Allow to use a tuple of keys in ``Table.sort``.  [#4671]
 
+  - Added interface to create ``Table`` directly from any "compatible" table
```
Maybe point out that the definition of "compatible" is described in more detail in the docs?  At first I was like "what does compatible mean?  and why is it in quotes?"

eteq commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+
+In order to improve interoperability between different table classes, an
+astropy |Table| object can be created directly from any other table
+object ``data`` that provides an ``__astropy_table__`` method.  In this case the
```
``object ` `data` ` that `` -> ``object that ... method (in the example below that object is named ` `data` `). ``

eteq commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,71 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+  def __astropy_table__(self, cls, copy, **kwargs):
+
+Your class implementation of this should include the ``**kwargs`` keyword
+args catch.  This is not currently used, but in the future additional
```
"args" -. "arguments" or maybe "the ``**kwargs`` technique of catching keyword arguments" or similar.

eteq commented:
@taldcroft - ok, I think I'm happy with this aside from the wordsmithing and @mhvk's point that the exact usage of the word ``copy`` seems counter-intuitive (the fix could be changing which way it goes, or a more verbose name, or both)

taldcroft commented:
One interesting point from https://github.com/lsst/afw/pull/64/ is that the current interface is not *quite* enough.  When converting an LSST table they have another kwarg `unviewable` that controls how the conversion is done (and whether an exception gets raised in some circumstances).

So this leads to the question of adding a `**kwargs` to the end of the `Table` arg list and passing that straight into the `**kwargs` that @mhvk presciently suggested we add to `__astropy_table__`.  :smile:  This certainly provides opportunity for flexibility that I had not considered before imagining adding `**kwargs` to the Table API.  So I am not terribly opposed to this idea, but I'd like to hear thoughts.

astrofrog commented:
I can see the value of having ``**kwargs``, so that would be fine by me.

eteq commented:
Yes, that also seems reasonable to me, @taldcroft . 

The one question I'm not clear on, though, is what to do with the `kwargs` when `__astropy_table__` is *not* called?  Do you just raise an exception just like the one you normally get when you call a function with invalid keywords?  That seems the safest road.

taldcroft commented:
@eteq - yes, an exception will be raised if there are `kwargs` but without a table-like object conversion happening.

I changed nomenclature from "compatible table" to "table-like object".  I think this is more clear and fits existing usage of *-like object.

I think that I have addressed all comments and successfully added the `**kwargs` interface.  The example now illustrates usage with something akin to what LSST needs.

taldcroft commented on CHANGES.rst:
```diff
@@ -126,6 +126,9 @@ New Features
     using the ``astropy.table.default_notebook_table_class`` configuration
     item. [#4886]
+  - Added interface to create ``Table`` directly from any "compatible" table
+  - Added interface to create ``Table`` directly from any table-like object
```
I don't know if this was me or rebase, but I'll fix this after tests are passing.

taldcroft commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,96 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+If ``copy`` is ``True`` then the ``__astropy_table__`` method must ensure that
+a copy of the original data is returned.  If ``copy`` is ``False`` then a
+reference to the table data should returned if possible.  If it is not possible
+(e.g. the original data are in a Python list or must be other transformed in
```
`other` => `otherwise`

taldcroft commented on docs/table/construct_table.rst:
```diff
@@ -960,3 +964,96 @@ a `~astropy.table.Column` object with a ``unit`` attribute::
+
+Internally the following call will be made to ask the ``data`` object
+to return a representation of itself as an astropy |Table|, respecting
+the ``copy`` preference of the original call to ``Table()``::
```
`call to QTable` not `Table`

mhvk commented on astropy/table/table.py:
```diff
@@ -2564,12 +2579,17 @@ class QTable(Table):
+        Additional keyword args when converting table-like object
 
     """
     def __init__(self, data=None, masked=None, names=None, dtype=None,
```
You could remove this whole `__init__`, since all you do is call the super... (but only worry about this if you need to make another change anyway).

mhvk commented:
This looks good to me. I particularly like the rewritten documentation; very clear!

bsipocz commented on docs/table/construct_table.rst:
```diff
@@ -527,6 +527,10 @@ for the ``data`` argument.
     the rest of the rows, a ValueError will be thrown.
 
+**table-like object**
+    If another table-like object has a `__astropy_table__` method then
```
This needs to be in double ticks.

bsipocz commented:
I had one minor comment that should fix the sphinx build failure (both warnings reference the same line number, though I'm not sure this would get rid of the first one).

taldcroft commented:
Tests are passing now.  Good to go?

mhvk commented:
> Tests are passing now. Good to go?

Yes, merging!  Nice!
